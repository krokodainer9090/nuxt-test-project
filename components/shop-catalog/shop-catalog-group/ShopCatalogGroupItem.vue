<script>
export default {
  name: 'ShopCatalogGroupItem',

  props: {
    item: {
      type: Object,
      default: _ => ({})
    }
  },

  data: () => ({
    isPriceIncreased: false,
    isSamePrice: false,
    canSetClasses: false
  }),

  computed: {
    itemPriceClasses () {
      const { isPriceIncreased, isSamePrice, canSetClasses } = this;

      return {
        'shop-catalog-group-item__price-container_increased': canSetClasses && isPriceIncreased && !isSamePrice,
        'shop-catalog-group-item__price-container_decreased': canSetClasses && !isPriceIncreased && !isSamePrice
      };
    },

    itemFullName () {
      const { N, count } = this.item;

      return N && count
        ? `${N}(${count})`
        : '';
    },

    itemPrice () {
      return this.item?.price || '';
    }
  },

  watch: {
    itemPrice (newPrice, oldPrice) {
      this.canSetClasses = true;

      this.isSamePrice = newPrice === oldPrice;
      this.isPriceIncreased = newPrice > oldPrice;
    }
  },

  methods: {
    clickItem () {
      this.$emit('on-click-item', this.item);
    }
  }
};
</script>

<template>
  <div class="shop-catalog-group-item">
    <div
      class="shop-catalog-group-item__content"
      @click="clickItem"
    >
      <span
        v-if="itemFullName"
        class="shop-catalog-group-item__name"
      >
        {{ itemFullName }}
      </span>
      <div
        :class="itemPriceClasses"
        class="shop-catalog-group-item__price-container"
      >
        <span
          v-if="itemPrice"
          class="shop-catalog-group-item__price"
        >
          {{ item.price }}
        </span>
      </div>
    </div>
  </div>
</template>

<style lang="scss">
@import "../../../styles/components/shop-catalog/shop-catalog-group/shop-catalog-group-item";
</style>
