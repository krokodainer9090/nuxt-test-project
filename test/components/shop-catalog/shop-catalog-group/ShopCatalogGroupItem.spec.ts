// Components
import ShopCatalogGroupItem from '~/components/shop-catalog/shop-catalog-group/ShopCatalogGroupItem.vue';

// Utilities
import { shallowMount, Wrapper } from '@vue/test-utils';
import { shopCatalogItem } from '~/test/utils/misc/constants';

// Constants
const itemContentClass = '.shop-catalog-group-item__content';

describe('ShopCatalogGroupItem', () => {
  const mountFunction: (options?: Object) => Wrapper<any> = (options = {}) => {
    return shallowMount(ShopCatalogGroupItem, {
      ...options
    });
  };

  it('should render empty item', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render item', async () => {
    const wrapper = mountFunction({
      propsData: {
        item: shopCatalogItem
      }
    });
    const emitObject = wrapper.emitted();

    expect(wrapper.html()).toMatchSnapshot();

    // Test emit on-click-item
    await wrapper.find(itemContentClass).trigger('click');

    expect(emitObject['on-click-item']?.[0]).toEqual([shopCatalogItem]);
  });
});
