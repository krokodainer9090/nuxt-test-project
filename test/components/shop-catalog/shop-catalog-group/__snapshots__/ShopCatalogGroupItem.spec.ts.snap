// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCatalogGroupItem should render empty item 1`] = `
"<div class=\\"shop-catalog-group-item\\">
  <div class=\\"shop-catalog-group-item__content\\">
    <!---->
    <div class=\\"shop-catalog-group-item__price-container\\">
      <!---->
    </div>
  </div>
</div>"
`;

exports[`ShopCatalogGroupItem should render item 1`] = `
"<div class=\\"shop-catalog-group-item\\">
  <div class=\\"shop-catalog-group-item__content\\"><span class=\\"shop-catalog-group-item__name\\">
      Суп с пекинской капустой(15)
    </span>
    <div class=\\"shop-catalog-group-item__price-container\\"><span class=\\"shop-catalog-group-item__price\\">
        42.8
      </span></div>
  </div>
</div>"
`;
